zebra-state/src/service/queued_blocks.rs

```diff
@@ -209,3 +209,52 @@ impl QueuedBlocks {
         self.blocks.drain()
     }
 }
+
+#[derive(Debug, Default)]
+pub(crate) struct SentHashes {
+    /// A growable ring buf to be used for pruning outdated hashes that are mostly
+    /// inserted in the order of their height
+    buf: VecDeque<(block::Hash, block::Height)>,
+
+    /// Stores a set of hashes that have been sent to the block write task but
+    /// may not be in the finalized state yet.
+    set: HashSet<block::Hash>,
+}
+
+impl SentHashes {
+    /// Inserts the hash into the set of sent hashes, and pushes the (hash, height)
+    /// onto the buf to be used for pruning outdated hashes.
+    ///
+    /// Assumes that hashes are mostly added in the order of their height for efficient pruning.
+    //
+    // Trades off some memory usage to avoid iterating over the entire set when non-finalized blocks
+    // are sent to the block write task, as would be the case with a HashMap.
+    pub fn add_with_height(&mut self, hash: block::Hash, height: block::Height) {
+        self.buf.push_back((hash, height));
+        self.set.insert(hash);
+    }
+
+    /// Iterates over the buf removing hashes from the SentHashes until reaching
+    /// the first hash with a height above the `height_bound`
+    ///
+    /// Does not guarantee that all hashes below the `height_bound` will be removed.
+    pub fn prune_by_height(&mut self, height_bound: block::Height) {
+        while let Some((hash, height)) = self.buf.pop_front() {
+            if height > height_bound {
+                self.buf.push_front((hash, height));
+                break;
+            } else {
+                self.set.remove(&hash);
+            }
+        }
+
+        // shrink to a reasonable cap in cases where there are many hashes that
+        // are not in order of their heights in the buf
+        self.buf.shrink_to(127);
```

Comment:
Nitpick: We prefer documented constants to integer literals, for code readability.

Reply:
There are two alternative fixes here:
- use `shrink_ to_fit()`
- don't shrink at all, and just keep using the extra memory - this is more efficient when we're syncing, because we don't reallocate, but less efficient when we've reached the tip

The amount of extra memory is about 75 kB:
`(1200 look ahead limit - 100 non-finalised chain) * (32 byte hash * 2 + 4 byte height)`

Whatever we do, we should do it to the `HashSet` as well.

Reply:
Nitpick: Whatever we do here, we should do it to the `HashSet` as well.

Reply:
Oops. I updated it to call `shrink_to_fit` on the `HashSet` in [`70beeb0`](https://github.com/ZcashFoundation/zebra/commit/70beeb07459d35940e4806342bba8daf48816d9c).